cosmos-tx/tests/integration.rs

```diff
@@ -0,0 +1,191 @@
+const ACCOUNT_PREFIX: &str = "cosmos";
+
+/// Denom name
+const DENOM: &str = "samoleans";
+
+/// Example memo
+const MEMO: &str = "test memo";
+
+#[test]
+fn msg_send() {
+    let sender_private_key = SigningKey::random();
+    let sender_account_id = sender_private_key
+        .public_key()
+        .account_id(ACCOUNT_PREFIX)
+        .unwrap();
+
+    let recipient_private_key = SigningKey::random();
+    let recipient_account_id = recipient_private_key
+        .public_key()
+        .account_id(ACCOUNT_PREFIX)
+        .unwrap();
+
+    let amount = Coin {
+        amount: 1u8.into(),
+        denom: DENOM.parse().unwrap(),
+    };
+
+    let msg_send = MsgSend {
+        from_address: sender_account_id.clone(),
+        to_address: recipient_account_id,
+        amount: vec![amount.clone()],
+    }
+    .to_msg()
+    .unwrap();
+
+    let chain_id = CHAIN_ID.parse().unwrap();
+    let sequence_number = 0;
+    let gas = 100_000;
+    let fee = Fee::from_amount_and_gas(amount, gas);
+    let timeout_height = 9001u16;
+    let tx_body = tx::Body::new(vec![msg_send], MEMO, timeout_height);
+
+    let tx = Builder::new(chain_id, ACCOUNT_NUMBER)
+        .sign_tx(tx_body, &sender_private_key, sequence_number, fee)
+        .unwrap();
+
+    let docker_args = [
+        "-d",
+        "-p",
+        &format!("{}:{}", RPC_PORT, RPC_PORT),
+        DOCKER_IMAGE,
+        CHAIN_ID,
+        &sender_account_id.to_string(),
+    ];
+
+    docker_run(&docker_args, || {
+        init_tokio_runtime().block_on(async {
+            let rpc_address = format!("http://localhost:{}", RPC_PORT);
+            let rpc_client = rpc::HttpClient::new(rpc_address.as_str()).unwrap();
+
+            rpc_client
+                .wait_until_healthy(Duration::from_secs(5))
+                .await
+                .unwrap();
+
+            // Workaround for what appears to be a Tendermint race condition.
+            // There is a short period after the RPC health check succeeds
+            // and the `/genesis` endpoint returns valid data where
+            // transactions fail due to an internal SDK panic in the
+            // "undefined" code space:
+            //
+            //     TxResult {
+            //         code: Err(111222), data: None, log: Log("panic"), info: Info(""),
+            //         gas_wanted: Gas(0), gas_used: Gas(0), events: [],
+            //         codespace: Codespace("undefined")
+            //     }
+            //
+            // This should ideally get fixed upstream, however there is
+            // presently no tracking issue for this bug.
+            //
+            // If we see it pop up locally or in CI, perhaps we could find
+            // some way of polling readiness via RPC, or retrying when this
+            // specific error is encountered.
+            // TODO(tarcieri): open upstream issue and remove this hack when fixed
+            tokio::time::sleep(Duration::from_secs(2)).await;
```

Comment:
This is probably the most suspect part of this PR.

It seems like polling an RPC endpoint to determine when the first block has been produced might help.

Reply:
Replaced this `sleep` with a `wait_for_first_block` function that polls the `/latest_block` endpoint and waits for a valid result.